ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -1519,16 +1541,26 @@ predicate parameterMatch(ParameterPosition ppos, ArgumentPosition apos) {
   (ppos.isHashSplat() or ppos.isSynthHashSplat()) and
   (apos.isHashSplat() or apos.isSynthHashSplat())
   or
-  exists(int pos |
+  exists(int pos, boolean hasActualSplatParam, boolean hasActualSplatArg |
     (
-      ppos.isSplat(pos)
+      ppos.isSplat(pos) and
+      hasActualSplatParam = true // allow matching with synthetic splat argument
       or
-      ppos.isSynthSplat() and pos = 0
+      ppos.isSynthSplat(hasActualSplatParam) and
+      pos = 0 and
+      // prevent synthetic splat parameters from matching synthetic splat arguments
+      // when direct positional matching is possible
+      (
+        hasActualSplatParam = true
+        or
+        hasActualSplatArg = true
+      )
     ) and
     (
-      apos.isSplat(pos)
+      apos.isSplat(pos) and
+      hasActualSplatArg = true // allow matching with synthetic splat parameter
```

Comment:
The comment seems misleading. It only actually matches a synthetic splat parameter if `pos=0` due to the restriction on line 1550. I continue to be confused by this clause 😕